File: backend/app/schemas/assignment.py
from pydantic import BaseModel, Field, validator
from datetime import datetime
from typing import Optional, List
from app.models.assignment import AssignmentStatus

class AssignmentCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255, description="Titolo del compito")
    description: str = Field(..., min_length=1, description="Descrizione del compito")
    instructions: str = Field(..., min_length=1, description="Istruzioni dettagliate per il compito")
    subject: str = Field(..., min_length=1, max_length=100, description="Materia del compito")
    due_date: datetime = Field(..., description="Data e ora di scadenza")
    points: int = Field(100, ge=1, le=1000, description="Punti massimi del compito")
    student_id: int = Field(..., description="ID dello studente")
    is_published: bool = Field(False, description="Se il compito è pubblicato")

    @validator('due_date')
    def validate_due_date(cls, v):
        if v <= datetime.utcnow():
            raise ValueError('La data di scadenza deve essere nel futuro')
        return v

class AssignmentUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, min_length=1)
    instructions: Optional[str] = Field(None, min_length=1)
    subject: Optional[str] = Field(None, min_length=1, max_length=100)
    due_date: Optional[datetime] = None
    points: Optional[int] = Field(None, ge=1, le=1000)
    is_published: Optional[bool] = None

class AssignmentResponse(BaseModel):
    id: int
    title: str
    description: str
    instructions: str
    subject: str
    due_date: datetime
    points: int
    is_published: bool
    created_at: datetime
    updated_at: datetime
    tutor_name: str
    student_name: str
    has_submission: bool = False
    submission_status: Optional[AssignmentStatus] = None
    submission_grade: Optional[int] = None

    class Config:
        from_attributes = True

class AssignmentListResponse(BaseModel):
    assignments: List[AssignmentResponse]
    total: int
    page: int
    per_page: int

class AssignmentSubmissionCreate(BaseModel):
    content: str = Field(..., min_length=1, description="Contenuto della consegna")
    assignment_id: int = Field(..., description="ID del compito")

class AssignmentSubmissionUpdate(BaseModel):
    content: Optional[str] = Field(None, min_length=1)

class AssignmentSubmissionResponse(BaseModel):
    id: int
    content: str
    status: AssignmentStatus
    grade: Optional[int] = None
    feedback: Optional[str] = None
    submitted_at: Optional[datetime] = None
    graded_at: Optional[datetime] = None
    created_at: datetime
    assignment_title: str
    student_name: str

    class Config:
        from_attributes = True

class AssignmentGrading(BaseModel):
    grade: int = Field(..., ge=0, le=100, description="Voto da 0 a 100")
    feedback: str = Field(..., min_length=1, description="Feedback per lo studente")

File: frontend/src/api/assignment.ts
import { apiClient } from './client';

export interface AssignmentCreate {
  title: string;
  description: string;
  instructions: string;
  subject: string;
  due_date: string; // ISO string
  points: number;
  is_published: boolean;
  student_id: number;
}

export interface AssignmentResponse {
  id: number;
  title: string;
  description: string;
  instructions: string;
  subject: string;
  due_date: string;
  points: number;
  is_published: boolean;
  created_at: string;
  updated_at: string;
  tutor_name: string;
  student_name: string;
  has_submission: boolean;
  submission_status?: string;
  submission_grade?: number;
}

export const assignmentApi = {
  // Crea un nuovo compito
  create: async (assignmentData: AssignmentCreate): Promise<AssignmentResponse> => {
    console.log('📝 [AssignmentAPI] Creazione compito:', assignmentData);
    const response = await apiClient.post('/assignments/', assignmentData);
    console.log('✅ [AssignmentAPI] Compito creato:', response.data);
    return response.data;
  },

  // Ottieni i compiti per tutor
  getTutorAssignments: async (): Promise<AssignmentResponse[]> => {
    console.log('📝 [AssignmentAPI] Recupero compiti tutor');
    const response = await apiClient.get('/assignments/tutor');
    console.log('✅ [AssignmentAPI] Compiti tutor ricevuti:', response.data);
    return response.data || [];
  },

  // Ottieni i compiti per studente
  getStudentAssignments: async (): Promise<AssignmentResponse[]> => {
    console.log('📝 [AssignmentAPI] Recupero compiti studente');
    const response = await apiClient.get('/assignments/student');
    console.log('✅ [AssignmentAPI] Compiti studente ricevuti:', response.data);
    return response.data || [];
  },

  // Ottieni un compito specifico
  getById: async (assignmentId: number): Promise<AssignmentResponse> => {
    console.log('📝 [AssignmentAPI] Recupero compito:', assignmentId);
    const response = await apiClient.get(`/assignments/${assignmentId}`);
    console.log('✅ [AssignmentAPI] Compito ricevuto:', response.data);
    return response.data;
  },

  // Aggiorna un compito
  update: async (assignmentId: number, assignmentData: Partial<AssignmentCreate>): Promise<AssignmentResponse> => {
    console.log('📝 [AssignmentAPI] Aggiornamento compito:', assignmentId, assignmentData);
    const response = await apiClient.put(`/assignments/${assignmentId}`, assignmentData);
    console.log('✅ [AssignmentAPI] Compito aggiornato:', response.data);
    return response.data;
  },

  // Elimina un compito
  delete: async (assignmentId: number): Promise<void> => {
    console.log('📝 [AssignmentAPI] Eliminazione compito:', assignmentId);
    await apiClient.delete(`/assignments/${assignmentId}`);
    console.log('✅ [AssignmentAPI] Compito eliminato');
  }
};
